### PR TITLE
[FrameworkBundle] framework.php_errors.log now accept a log level

### LINKED ISSUE #26267: Deprecated error are logged on production
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4.4

I have a `env.php` file with the following line to set the error reporting level:

```php
error_reporting(E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED & ~E_STRICT);
```

This file is dumped on the composer autoload thanks to this configuration:

```json
{
    "autoload": {
        "psr-4": { "": "src/" },
        "files": [
            "app/env.php",
            "app/AppKernel.php"
        ]
    },
    "autoload-dev": {
        "psr-4": {
            "Dev\\": "src-dev",
            "Tests\\": "tests/"
        },
        "files": [
            "app/env_dev.php",
            "vendor/symfony/symfony/src/Symfony/Component/VarDumper/Resources/functions/dump.php"
        ]
    },
}
```

And of course, the `composer dump` is called with `--no-dev` on the production server.

Currently, if a exception is thrown on production and deprecated error are triggered, they are logged on the production log file.

So I start with this tiny test:

```php
require_once __DIR__.'/vendor/autoload.php';

var_dump(error_reporting());

trigger_error('This is a deprecation.', E_USER_DEPRECATED);
trigger_error('This is an error.', E_USER_ERROR);
```

And here is the result:

```
$ php current/test.php 
int(6143)
PHP Fatal error:  This is an error. in /var/www/project/prod/releases/20180221183905/test.php on line 10
```

It works. The deprecated error is ignored and the error_reporting value is `6143`.

First, we can supposed the error_reporting is changed later. So I tried with this simple controller action:

```php
public function failAction(): void
{
    trigger_error('Error reporting: '.error_reporting(), E_USER_DEPRECATED);
    throw new \Exception('Test exception with error reporting: '.error_reporting());
}
```

After a call of this action, I have the following log:

```
$ grep -B1 'Test exception' shared/var/logs/prod-2018-02.log 
[2018-02-22 11:16:37] php.INFO: User Deprecated: Error reporting: 6143 {"exception":"[object] (ErrorException(code: 0): User Deprecated: Error reporting: 6143 at /var/www/project/prod/releases/20180221183905/src/AppBundle/Controller/DefaultController.php:167)"} []
[2018-02-22 11:16:37] request.CRITICAL: Uncaught PHP Exception Exception: "Test exception with error reporting: 6143" at /var/www/project/prod/releases/20180221183905/src/AppBundle/Controller/DefaultController.php line 168 {"exception":"[object] (Exception(code: 0): Test exception with error reporting: 6143 at /var/www/project/prod/releases/20180221183905/src/AppBundle/Controller/DefaultController.php:168)"} []
```

As you can see, both deprecation and error are logged and the error_reporting value is still the same: `6143`

This is why I'm thinking about a bug, because the error reporting level does not change so the deprecation should not be reported.

If you think it is not, I'm curious to know what is the issue. :smile:

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #26267   
| License       | MIT
| Doc PR        |  todo 

We are testing that `framework.php_errors.log` is either a bool or an int (set the value of the log level you want).
This fixes #26267, so it gives a way to not log some level on production.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
would be great to turn the `framework.php_errors.log` config entry into a "bool|int", and when an "int" is provided, use the value to set the parameter

**Comment by Simperfit** [comment]
@nicolas-grekas I have changed this PR, could you please guide me on how to write test for this ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -601,10 +601,14 @@ private function registerDebugConfiguration(array $config, ContainerBuilder $con
 
         $definition = $container->findDefinition('debug.debug_handlers_listener');
 
-        if (!$config['log']) {
+        if (\is_bool($config['log']) && !$config['log']) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `if (!$config['log']) {` is enough: `0` also disables the logger

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -601,10 +601,14 @@ private function registerDebugConfiguration(array $config, ContainerBuilder $con
...
             $definition->replaceArgument(1, null);
         }
 
+        if (\is_int($config['log'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: if (\is_int($config['log']) && $config['log']) {

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -833,7 +833,7 @@ private function addPhpErrorsSection(ArrayNodeDefinition $rootNode)
                     ->info('PHP errors handling configuration')
                     ->addDefaultsIfNotSet()
                     ->children()
-                        ->booleanNode('log')
+                        ->scalarNode('log')
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: maybe forbid anything else than a bool|int?

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
for tests, please check the existing ones, I don't have specific ideas on the topic sorry

**Comment by Simperfit** [comment]
Status: Needs Review

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 3 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: done

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Simperfit** [comment]
Tests Fixed.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -833,10 +833,14 @@ private function addPhpErrorsSection(ArrayNodeDefinition $rootNode)
...
                             ->info('Use the app logger instead of the PHP logger for logging PHP errors.')
                             ->defaultValue($this->debug)
                             ->treatNullLike($this->debug)
+                            ->validate()
+                            ->ifTrue(function ($v) { return !(\is_int($v) || \is_bool($v)); })
+                            ->thenInvalid('The "php_errors.log" parameter should be either a int or a boolean.')
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: `either an integer or a boolean`?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 4 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: @fabpot updated.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -833,10 +833,14 @@ private function addPhpErrorsSection(ArrayNodeDefinition $rootNode)
...
                     ->addDefaultsIfNotSet()
                     ->children()
-                        ->booleanNode('log')
+                        ->scalarNode('log')
                             ->info('Use the app logger instead of the PHP logger for logging PHP errors.')
                             ->defaultValue($this->debug)
                             ->treatNullLike($this->debug)
+                            ->validate()
+                            ->ifTrue(function ($v) { return !(\is_int($v) || \is_bool($v)); })
+                            ->thenInvalid('The "php_errors.log" parameter should be either a integer or a boolean.')
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: an integer :)

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 5 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: This is good now :p.

**Comment by fabpot** [comment]
Thank you @Simperfit.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -833,10 +833,14 @@ private function addPhpErrorsSection(ArrayNodeDefinition $rootNode)
                     ->info('PHP errors handling configuration')
                     ->addDefaultsIfNotSet()
                     ->children()
-                        ->booleanNode('log')
+                        ->scalarNode('log')
                             ->info('Use the app logger instead of the PHP logger for logging PHP errors.')
```
Comment (bendavies) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: is this really descriptive now?

**Comment by mathieu-pousse** [comment]
Hi, 

This PR looks great to me, but I am not sure to ensure the aim.
Does  the log level behind the scene correspond to something like 'E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED & ~E_STRICT' or to DEVUG, INFO, WARN, ?

In any case, I think this should be explain in the documentation

**Comment by SublimeRaiser** [comment]
> This PR looks great to me, but I am not sure to ensure the aim.
> Does the log level behind the scene correspond to something like 'E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED & ~E_STRICT' or to DEVUG, INFO, WARN, ?
> 
> In any case, I think this should be explain in the documentation

Here is a link to the docs: [https://symfony.com/doc/current/reference/configuration/framework.html#log](https://symfony.com/doc/current/reference/configuration/framework.html#log)

**Pull request merged**